### PR TITLE
Driver `/solve` functionality part 3

### PR DESCRIPTION
**Pull request by sistemd**
Progress on https://github.com/cowprotocol/services/issues/895. Fixes https://github.com/cowprotocol/services/issues/914.

- Implements multiplexing for the solvers.
- Introduces a `boundary` module to serve as an [anti-corruption layer](https://gist.github.com/ennmichael/0f68a4e0c33df80f9d415b20f9848bcf) to integrate with the rest of the system. I did this because I realized that re-implementing scoring is error prone, and re-implementing all of liquidity collection is simply too much work to do at once. (Also, @MartinquaXD showed me [the cow-native-liquidity repo](https://github.com/cowprotocol/cow-native-liquidity) which seems to have been an initial effort to factor out liquidity collection into its own service. I feel like that's a better direction to take in the future.)
- Introduces a rough sketch for what the  integration will look like, what the responsibilities of `driver` will be, and what code the `driver` will rely on from `shared` for now.
- Sketches the interface for settlement simulation.
- Implements the logic around approvals and allowances.
- Implements part of the settlement encoding process (this is at the boundary); the next PR's main focus will be on finishing this.
- Replaces `ethereum_types` with `primitive_types`, I should have probably done this from the start.

### CODE REVIEW

File: crates/driver/src/api/mod.rs
Comment (nlordell) [reply crates/driver/src/api/mod.rs:51]: Agree, I would call the loop multiplexing, and this would conceptually be "mounting" to me.

File: crates/driver/src/boundary/mod.rs
Comment (nlordell) [inline crates/driver/src/boundary/mod.rs:1]: 😂

File: crates/driver/src/node/mod.rs
Comment (nlordell) [inline crates/driver/src/node/mod.rs:88]: I think this is a cool abstraction!

File: crates/driver/src/node/mod.rs
Comment (nlordell) [inline crates/driver/src/node/mod.rs:21]: I'm not super fond of this, as to me this doesn't represent an "Ethereum node" but more "Blockchain state".

File: crates/driver/src/simulator/mod.rs
Comment (nlordell) [inline crates/driver/src/simulator/mod.rs:55]: For the future, and not here - I would actually think that:
- Tenderly is just as fast, and maybe even faster with the `fast` flag
- Tenderly simulations allow for a "save_on_failure" flag to be set, so we automatically store failed simulations in the project's dashboard which would be **amazing** for debugging.

I think, for this PR it is fine to keep it like this, but as an additional "nice to have" feature, we should support simulations on Tenderly as well.

**Review by nlordell — COMMENTED** [review]
Started reviewing, didn't get through everything yet.

File: crates/driver/src/simulator/mod.rs
Comment (MartinquaXD) [reply crates/driver/src/simulator/mod.rs:55]: Wow, didn't know that `tenderly` could be faster than using the node for simulation (seems like more overhead). But I guess it makes sense. They probably have their whole infrastructure optimized for running simulations.

File: crates/driver/src/logic/competition/solution/settlement.rs
Comment (MartinquaXD) [reply crates/driver/src/logic/competition/solution/settlement.rs:71]: Just wanted to mention it because it can be quite annoying to move code from `sync` to `async` because it's so fundamentally different.

File: crates/driver/src/logic/competition/solution/settlement.rs
Comment (nlordell) [reply crates/driver/src/logic/competition/solution/settlement.rs:28]: Personally, when I looked through the function, it looks like it is taking an "abstract settlement" from the solver (so a settlement containing things like "please approve this much" which may require fetching on-chain state) and turning it into an "encod**able** settlement" that can be easily transformed into transaction calldata (so the "a solution encoded in such a way to be ready for publishing on-chain" that @ennmichael described).

In that sense, this function feels like it is doing one thing only: taking an "abstract settlement" from the solvers and turning it into something that can be concretely used for generating calldata.

File: crates/driver/src/logic/competition/solution/settlement.rs
Comment (nlordell) [reply crates/driver/src/logic/competition/solution/settlement.rs:47]: I think it may also need order data (which is included in the auction, but is not sent in complete form to the solvers - things like `appData` have no bearing on how an order executes, but is required for encoding settlement calldata because of the signature verification).

So I do think `auction` is needed here as well.

File: crates/driver/src/node/mod.rs
Comment (nlordell) [reply crates/driver/src/node/mod.rs:21]: I would call it `Chain` or `Blockchain` if anything - just so we are more chain agnostic in our naming.

`Eth` is fine as well.

File: crates/driver/src/simulator/mod.rs
Comment (nlordell) [reply crates/driver/src/simulator/mod.rs:55]: It isn't with the "full" flag - (since it does deeper tracing, and some source mapping), but the "fast" flag skips this and only does the simulation with the trace (which is a pretty small overhead in general - most nodes are IO bound on state lookup which Tenderly implements very well from what I understand - and anecdotally experienced).

Comment (nlordell) [reply crates/driver/src/simulator/mod.rs:55]: Either way, even it it is slower, it should be in the same order of magnitude, and since this simulation is done once I think the advantages (more convenient inspection in the Tenderly dashboard) completely outweigh the negative.

File: crates/driver/src/boundary/mod.rs
Comment (nlordell) [reply crates/driver/src/boundary/mod.rs:1]: Name is good and descriptive. Just wanted to share the laugh 😛

File: crates/driver/src/logic/eth.rs
Comment (nlordell) [reply crates/driver/src/logic/eth.rs:45]: Personally, I am a huge fan of these new-types. I believe they simultaneously help in two ways:

1. convey code intent more clearly. For example, in the code below, you very clearly know you are talking about buy and sell tokens and not amounts because of their type:
    ```rust
    let sell: Token;
    let buy: Token;
    ```
    
2. compile-time hints for hard to catch bugs. For example, a pretty innocuous:
    ```rust
    async fn get_balance(owner: Address, token: Address) -> Result<U256>;
    ```

    Now, at a glance, without looking back up to the signature, it is not obvious which one of these is correct:
    ```rust
    get_balance(my_wallet_address, cow_token).await;
    get_balance(cow_token, my_wallet_address).await;
    ```

The downside to me is additional boilerplate, but I think it is a very small price to pay in the long run.

File: crates/driver/src/logic/competition/solution/approvals.rs
Comment (nlordell) [reply crates/driver/src/logic/competition/solution/approvals.rs:16]: The thing that tripped me up here is that `Approval` is an "approval request" indicating that "we need at least X amount of tokens approved for spender Y" and not the approval that should be set with an approval call.

Essentially, this turns into a:
```rust
if get_allowance(approval.spender) < approval.amount {
    Some(build_approval_interaction(approval.spender) // just spender, amount is U256::MAX
} else {
    None
}
```

And not:
```
build_approval_interaction(approval) // amount + spender
```

I would still not panic here because this condition could happen because of some **external** "invalid" data and not, for example, a logic bug. I would reserve the use of `panic`s for "this should never happen" and not "this can happen if an external system misbehaves".

Comment (nlordell) [reply crates/driver/src/logic/competition/solution/approvals.rs:16]: Maybe we should rename this to a `RequiredAllowance(s)` since it represents the minimum allowance that is needed and not the approvals that need to be made.

File: crates/driver/src/boundary/mod.rs
Comment (nlordell) [inline crates/driver/src/boundary/mod.rs:7]: This gist is short enough that it is worth inlining IMO.

**Review by nlordell — COMMENTED** [review]
Code looks great. Two comments:
- I think `EthNode` should be renamed
- I think the `Approvals` needs to be slightly reworked to not panic and I also think it should be renamed as well (as it isn't a list of approvals to make, but rather a list of required allowances that need to be satisfied).

File: crates/driver/src/logic/competition/solution/approvals.rs
Comment (nlordell) [reply crates/driver/src/logic/competition/solution/approvals.rs:16]: > Why do the solvers return the approvals in /solve anyway? Isn't this something that the driver can determine based on orders? And furthermore, to make the lives of solvers easier, isn't this something that the driver should do?

The solvers propose `interactions`, which are these "block-box" calls to external contracts. Typically, when used for swapping some tokens against an on-chain AMM, the interaction will only work if there is the required approval given by the Settlement contract to some address specific to the on-chain AMM. In that regard, the approvals are not needed for users and orders (so, a settlement with perfect CoWs only including user orders will never need additional approvals).

Note that we include `inputs` and `outputs` with interactions, so in theory, we know the tokens and the amounts for the approvals that are needed. Unfortunately, we don't know the target. Some AMMs require approval directly on the contract that you interact for trading (Uniswap to their router contract for example), while others, require approvals to a completely different contract (ParaSwap swapper contract has a separate token approvals target that you need to set an allowance for).

> `RequiredAllowance` concept feels slightly unnatural

Interesting, maybe I've been in the space for so long, but this is a pretty "typical" concept to me. For example [0x includes this information in their swap response]() and [1Inch has a dedicated route to give you this information]().

One thing that we can do to make this more natural, is to attach an `approval_target: Address` to each `Interaction` or (since it needs has [input tokens and amounts](https://github.com/cowprotocol/services/blob/f87678d78cd431056e5b9ba2d1130ca41168b785/crates/shared/src/http_solver/model.rs#L150-L155)). This way it is closer to where it is actually needed (i.e. it is a pre-req for an interaction) and de-duplicates some of the data (token and amounts).

Comment (nlordell) [reply crates/driver/src/logic/competition/solution/approvals.rs]: Nice - this looks good.

We have a list of "Required" allowances from the solvers. These get normalized and summed up. Then these get compare to "Existing" allowances that have already been set, to generate the list of "Approval" transactions that need to be made.

I think this makes a bunch of sense and love the new naming! 💪

**Review by nlordell — APPROVED** [review]
Code looks good.

Only open suggestion from me is the `EthNode` rename, but it wouldn't block the merge (and we are already discussing alternatives).

File: crates/driver/src/simulator/mod.rs
Comment (MartinquaXD) [reply crates/driver/src/simulator/mod.rs:35]: I think the drawbacks you outlines regarding `dyn` and generics are definitely valid and the approach you chose to avoid both definitely seems appealing. I just have a few gripes with that:
1. IMO it would be great if we could reduce the number of concepts we use in the code base to a minimum and agree on a minimum set of them. In this case your approach looks like it could generally replace the usage of `dyn` in our code base.
2. If we would like to replace the usage of `dyn` with this pattern I'm a bit worried that this pattern would look quite nasty for cases where there are lots of types that implement a trait (e.g. `PriceEstimating`).

I looked a bit around and it seems enough people share my concern about maintenance that someone created a crate that takes care of all the boiler plate code and makes it look like the `dyn` code we have right now where we implement traits on structs in stead of the enum itself.
I reimplemented the gist of the `simulator` logic with the [`enum_dispatch`](https://docs.rs/enum_dispatch/latest/enum_dispatch/) crate and it looks reasonable enough to me that we could liberally replace the current usage of `Box<dyn MyTrait>` with the new concept.

<details>
<summary>with enum_dispatch</summary>

```rust
#![allow(dead_code)]
use async_trait::async_trait;
use enum_dispatch::enum_dispatch;

/// crates/driver/src/simulator/mod.rs
#[async_trait]
#[enum_dispatch(Simulator)]
trait Simulating {
    async fn simulate(&self, _settlement: &[u8]);
}

#[enum_dispatch(Simulator)]
#[derive(Debug, Clone)]
enum Simulator {
    Node,
    Tenderly,
}

/// crates/driver/src/simulator/node.rs
#[derive(Debug, Clone)]
struct Node {
    url: String,
}

impl Node {
    pub fn new(url: String) -> Self {
        Self { url }
    }
}

#[async_trait]
impl Simulating for Node {
    async fn simulate(&self, _settlement: &[u8]) {
        println!("simulated with node");
    }
}

/// crates/driver/src/simulator/tenderly.rs
#[derive(Debug, Clone)]
struct Tenderly {
    url: String,
    key: String,
}

impl Tenderly {
    pub fn new(url: String, key: String) -> Self {
        Self { url, key }
    }
}

#[async_trait]
impl Simulating for Tenderly {
    async fn simulate(&self, _settlement: &[u8]) {
        println!("simulated with tenderly");
    }
}

/// crates/driver/src/main.rs
#[tokio::main]
async fn main() {
    let simulators = vec![
        Simulator::Node(Node::new("node".into())),
        Simulator::Tenderly(Tenderly::new("a".into(), "b".into())),
    ];
    let settlement = vec![1, 2, 3, 4, 5, 6];
    for simulator in &simulators {
        simulator.simulate(&settlement).await;
    }
}
```
</details>

File: crates/driver/src/logic/competition/solution/approvals.rs
Comment (MartinquaXD) [reply crates/driver/src/logic/competition/solution/approvals.rs]: Yeah, I'm not suggesting to change the code. Just wanted to point this out since you recently asked about the usage of `BTreeMap`.

**Review by vkgnosis — APPROVED** [review]
Structure of this looks fine. I don't see anything obviously wrong. Can't review completely yet because there many open tasks but we can merge to move forward with it.

**Pull request merged**
